File: centwise/apps/android/app/src/main/kotlin/com/centwise/data/repository/SmartRulesRepository.kt
package com.centwise.data.repository

import com.centwise.data.models.RuleMatchType
import com.centwise.data.models.SmartRule
import com.centwise.data.models.TransactionType
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Repository for managing user-defined Smart Rules.
 * Smart Rules allow automatic categorization of transactions based on merchant keywords.
 */
class SmartRulesRepository private constructor() {

    private val _rules = MutableStateFlow<List<SmartRule>>(
        listOf(
            SmartRule(
                name = "Foodpanda is Food",
                keyword = "Foodpanda",
                matchType = RuleMatchType.CONTAINS,
                categoryName = "Food & Dining",
                transactionType = TransactionType.EXPENSE,
                isEnabled = true
            ),
            SmartRule(
                name = "Pathao is Transport",
                keyword = "Pathao",
                matchType = RuleMatchType.CONTAINS,
                categoryName = "Transport",
                transactionType = TransactionType.EXPENSE,
                isEnabled = true
            ),
            SmartRule(
                name = "Daraz is Shopping",
                keyword = "Daraz",
                matchType = RuleMatchType.CONTAINS,
                categoryName = "Shopping",
                transactionType = TransactionType.EXPENSE,
                isEnabled = true
            ),
            SmartRule(
                name = "Chaldal is Groceries",
                keyword = "Chaldal",
                matchType = RuleMatchType.CONTAINS,
                categoryName = "Groceries",
                transactionType = TransactionType.EXPENSE,
                isEnabled = true
            ),
            SmartRule(
                name = "Skitto is Mobile Recharge",
                keyword = "Skitto",
                matchType = RuleMatchType.CONTAINS,
                categoryName = "Mobile Recharge",
                transactionType = TransactionType.EXPENSE,
                isEnabled = true
            )
        )
    )

    val rules: StateFlow<List<SmartRule>> = _rules.asStateFlow()

    fun addRule(rule: SmartRule) {
        _rules.value = listOf(rule) + _rules.value
    }

    fun updateRule(rule: SmartRule) {
        _rules.value = _rules.value.map { if (it.id == rule.id) rule else it }
    }

    fun deleteRule(id: String) {
        _rules.value = _rules.value.filter { it.id != id }
    }

    fun toggleRule(id: String, isEnabled: Boolean) {
        _rules.value = _rules.value.map {
            if (it.id == id) it.copy(isEnabled = isEnabled) else it
        }
    }

    /**
     * Checks all enabled Smart Rules against the merchant/party string.
     * Returns the first matching rule, or null if no rule applies.
     */
    fun applyRules(merchantOrParty: String): SmartRule? {
        val enabledRules = _rules.value.filter { it.isEnabled }
        for (rule in enabledRules) {
            if (rule.matchType.matches(merchantOrParty, rule.keyword)) {
                return rule
            }
        }
        return null
    }

    companion object {
        val shared = SmartRulesRepository()
    }
}

File: centwise/apps/ios/Centwise/Data/Repositories/SmartRulesRepository.swift
import Foundation
import Combine

public final class SmartRulesRepository: ObservableObject {
    public static let shared = SmartRulesRepository()

    @Published public private(set) var rules: [SmartRule] = [
        SmartRule(
            name: "Foodpanda is Food",
            keyword: "Foodpanda",
            matchType: .contains,
            category: TransactionRepository.shared.category(id: "food"),
            transactionType: .expense,
            isEnabled: true
        ),
        SmartRule(
            name: "Pathao is Transport",
            keyword: "Pathao",
            matchType: .contains,
            category: TransactionRepository.shared.category(id: "transport"),
            transactionType: .expense,
            isEnabled: true
        ),
        SmartRule(
            name: "Daraz is Shopping",
            keyword: "Daraz",
            matchType: .contains,
            category: TransactionRepository.shared.category(id: "shopping"),
            transactionType: .expense,
            isEnabled: true
        ),
        SmartRule(
            name: "Chaldal is Groceries",
            keyword: "Chaldal",
            matchType: .contains,
            category: TransactionRepository.shared.category(id: "food"),
            transactionType: .expense,
            isEnabled: true
        )
    ]

    public init() {}

    public func addRule(_ rule: SmartRule) {
        rules.insert(rule, at: 0)
    }

    public func updateRule(_ rule: SmartRule) {
        if let index = rules.firstIndex(where: { $0.id == rule.id }) {
            rules[index] = rule
        }
    }

    public func deleteRule(id: String) {
        rules.removeAll { $0.id == id }
    }

    public func toggleRule(id: String, isEnabled: Bool) {
        if let index = rules.firstIndex(where: { $0.id == id }) {
            rules[index].isEnabled = isEnabled
        }
    }

    public func applyRules(merchantOrParty: String) -> SmartRule? {
        let enabled = rules.filter { $0.isEnabled }
        for rule in enabled {
            if rule.matchType.matches(merchant: merchantOrParty, keyword: rule.keyword) {
                return rule
            }
        }
        return nil
    }
}
